src/jit/morph.cpp

```diff
@@ -6940,263 +6945,236 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+                calleeFloatArgRegCount += numIntRegAndFloatRegArgs.second;
 
-                    calleeArgRegCount += size;
+                if ((callerIntArgRegCount > 1 || calleeFloatArgRegCount > 1) || ((callerIntArgRegCount + calleeFloatArgRegCount) > 0))
+                {
+                    // TODO-CQ: Fix LowerFastTailCall(GH #12644)
+                    //
+                    // LowerFastTailCall currently only support swapping/copying 1 slot
+                    // stack argument. All arguments are assumed to be 1 slot size.
+                    //
+                    // Because of this requirement, we lose the opportunity to fast tail call
+                    // >8 byte size structs on non windows if either the callee or caller
+                    // have stack arguments.
+                     hasLargerThanOneStackSlotSizedStruct = true;
                 }
-
-#elif defined(WINDOWS_AMD64_ABI)
-
-                ++calleeArgRegCount;
-
-#endif // UNIX_AMD64_ABI
-
-#else
-                assert(!"Target platform ABI rules regarding passing struct type args in registers");
-                unreached();
-#endif //_TARGET_AMD64_ || _TARGET_ARM64_
             }
-            else
+
+            if (arg->passedByRef)
             {
-                hasMultiByteStackArgs = true;
+                hasByrefStruct = true;
+                break;
             }
         }
-        else
-        {
-            varTypeIsFloating(argx) ? ++calleeFloatArgRegCount : ++calleeArgRegCount;
-        }
+    }
+
+    if (callee->HasRetBufArg()) // RetBuf
+    {
+        // We don't increment callerIntArgRegCount here, since it is already in callee->gtCallArgs.
 
-        // We can break early on multiByte cases.
-        if (hasMultiByteStackArgs)
+        // If callee has RetBuf param, caller too must have it.
+        // Otherwise go the slow route.
+        if (info.compRetBuffArg == BAD_VAR_NUM)
         {
-            break;
+            reportFastTailCallDecision("Callee has RetBuf but caller does not.", 0, 0);
+            return false;
         }
     }
 
     const unsigned maxRegArgs = MAX_REG_ARG;
+    const size_t callerStackSize = info.compArgStackSize;
 
-// If we reached here means that callee has only those argument types which can be passed in
-// a register and if passed on stack will occupy exactly one stack slot in out-going arg area.
-// If we are passing args on stack for the callee and it has more args passed on stack than
-// the caller, then fast tail call cannot be performed.
-//
-// Note that the GC'ness of on stack args need not match since the arg setup area is marked
-// as non-interruptible for fast tail calls.
+    // Do not support byref struct types. The analysis required to prove that the struct will not
+    // create an outgoing local to pass to the callee is too much work.
+    //
+    // For example, consider the following case:
+    //
+    //  (48 byte struct)
```

Comment:
Could you replace this diagram with legal C code, to be more clear? It took me a few moments to understand the shorthand